src/category_theory/limits/shapes/images.lean

```diff
@@ -273,3 +274,112 @@ end
 end
 
 end category_theory.limits
+
+namespace category_theory.limits
+
+variables {C : Type u} [𝒞 : category.{v} C]
+include 𝒞
+
+section
+
+instance {X Y : C} (f : X ⟶ Y) [has_image f] : has_image (arrow.mk f).hom :=
+show has_image f, by apply_instance
+
+end
+
+section has_image_map
+
+/-- An image map is a morphism `image f → image g` fitting into a commutative square and satisfying
+    the obvious commutativity conditions. -/
+class has_image_map {f g : arrow C} [has_image f.hom] [has_image g.hom] (sq : f ⟶ g) :=
+(map : image f.hom ⟶ image g.hom)
+(factor_map' : factor_thru_image f.hom ≫ map = sq.left ≫ factor_thru_image g.hom . obviously)
+(map_ι' : map ≫ image.ι g.hom = image.ι f.hom ≫ sq.right . obviously)
+
+restate_axiom has_image_map.factor_map'
+restate_axiom has_image_map.map_ι'
+attribute [simp, reassoc] has_image_map.factor_map has_image_map.map_ι
+
+variables {f g : arrow C} [has_image f.hom] [has_image g.hom] (sq : f ⟶ g)
+
+section
+local attribute [ext] has_image_map
```

Comment:
Any reason to make this `local`? It seems a reasonable lemma to have `ext` be able to use.

Reply:
Why would you use `ext` on a subsingleton?

Reply:
Because then `tidy` can make easy progress, because it call `ext`.

Reply:
It shouldn't call `ext`. There is no need to: `has_image_map` is a subsingleton, so if we have a goal `x = y` where `x y : has_image_map sq`, then  `tidy` will just close the goal with `exact dec_trivial`. It would be bad if `tidy` called `ext`, because then it couldn't automatically solve the goal any more.

Reply:
`tidy` will always call `dec_trivial` before trying `ext`, so it won't actually fall down this hole.

I guess the thing I hadn't appreciated is that applying this `ext` attribute creates what is actually a quite poor `ext` lemma (which gets named `has_image_map_ext`). Of course, the attribute `@[ext]` on this lemma disappears by the end of the section, so it's not particularly bad pollution.

I'm now happy to leave as is --- although if you'd prefer to not have the `has_image_map_ext` declaration be synthesised and left around at all I think that would be reasonable. (Just do the `cases ..., congr` yourself in the construction of the subsingleton.)